### DISCUSSION TITLE
are Nix flakes plausible for godot main repo?

### DISCUSSION THREAD

**Discussion by RusticCraftsman** [godotengine/godot-proposals#9334] [Platforms]
# What is a flake?
If you don't know what flakes are , a flake is technically a directory which cointains a flake.nix. Flakes are written in the nix lang and they can be used for building and running packages, having development isolated environments, all of this and more while ensuring reproducibility and granular control. ([more info ](https://zero-to-nix.com/concepts/flakes))

### I consider flakes a great adittion to any project, and here are a number of things you could do with flakes on the godot repo:

- Devshells for developing the engine
- Automated local building
- any version of Godot declaratively installed / runned down to git revision level
- easy integration of a specific Godot version to any project
- easy modification of the build process

All of this while being reproducible and declarative.

### If you come up with more, i encourage to comment on them in this discussion

# Now for the questions i have for you all :

**Comment by tkgalk** [comment]
> are Nix flakes plausible for godot main repo?

I don't know, it's your proposal. Are they?

**Reply by RusticCraftsman** [reply]
I would say that it is definetly possible and plausible to do, altough not worth it unless the devs reach a consensus that having and mantaining the flake is worth it, they would have to decide, it's their project after all.

**Comment by BirdeeHub** [comment]
Why? Godot is already on nixpkgs, you can make a different flake to do this every day of the week if you wanted. A dev shell could be nice I guess for godot contributors who use nix? But you wouldnt need a bunch of flakes, just a single flake.nix file with the package, a debug build of the package, and a dev shell with the build tools in it. But it is packaged on nixpkgs, so you can already just do this if you want, just override the source to newest and make a dev shell with the build tools in it. Could occasionally be a little involved though if you have to make build changes without its own flake, but usually even that is just doable with an override

Neovim has a contrib folder, and in it is a flake.nix with those things in it, maybe godot could do that, it could make some things nicer, but more isnt needed. https://github.com/neovim/neovim/blob/master/contrib/flake.nix <--- here it is. Maybe something like this could be done. It would make it easier to contribute to and use the *most* up to date version of godot. It also may improve CI. But more would be overkill.

But yes. It could 100% be done. It can be done for almost any existing open source software. The main issue is precompiled stuff with not available source. And uhhh, also the precompiled proprietary stuff that java sometimes shoves into fat jars (which can also be solved but its harder). Godot does not have this issue, nix is fair game.

**Reply by RusticCraftsman** [reply]
Well, it is already on nixpkgs, the versions provided could suffice your use case, or they could not:

for example the edge case of needing godot-mono version 4.3 dev 5 . Latest mono version provided by nixpkgs is 3.5.2 which is more than a year old, we could consider this a nixpkgs problem of course, but even if we got the latest stable package, you would have version 4.2.1 , which is more than 3 months old. 

Does this happen in real projects? yes, i was part of a project in the past that used dev versions and i know other projects using dev versions of godot.

Could not you use the binary on nixos and run it with something like steam-run or nix-ld ? Absolutely, but that defeats the porpouse and it's more prone to errors.

Could you make your own flake in your project solving your specific version problem? yeah, but that is not what im discussing here.

> Neovim has a contrib folder, and in it is a flake.nix with those things in it, maybe godot could do that, it could make some things nicer, but more isnt needed. https://github.com/neovim/neovim/blob/master/contrib/flake.nix <--- here it is. Maybe something like this could be done. It would make it easier to contribute to and use the most up to date version of godot. But more would be overkill.
> 
> But yes. It could 100% be done. It can be done for almost any existing open source software. The main issue is precompiled stuff with not available source. Godot does not have this issue, nix is fair game.

I agree

**Reply by BirdeeHub** [reply]
> for example the edge case of needing godot-mono version 4.3 dev 5 . Latest mono version provided by nixpkgs is 3.5.2 which is more than a year old, we could consider this a nixpkgs problem of course, but even if we got the latest stable package, you would have version 4.2.1 , which is more than 3 months old.


You can override the src attribute and possibly the build steps that change, but yeah for newest, there could be some issues, it definitely makes it harder to contribute if changes to the build process occurred between the latest nixpkgs version and master, at a certain point you would just end up copy pasting the nixpkgs source and modifying it manually XD

A flake.nix file like neovim has could definitely improve things, especially for those who wish to contribute to godot.

But yeah a good part of my point was, it wouldnt take a lot. a flake.nix file that outputs a normal and debug build of the package, and a dev shell with the build tools would be plenty sufficient. And it doesnt even need to be at root level of the repo, we can specify the directory of the flake. basically just 2 files (flake.nix and flake.lock) somewhere is all thats required

**Reply by RusticCraftsman** [reply]
> You can override the src attribute and possibly the build steps that change, but yeah for newest, there could be some issues, it definitely makes it harder to contribute if changes to the build process occurred between the latest nixpkgs version and master, at a certain point you would just end up copy pasting the nixpkgs source and modifying it manually XD

I can say it has changed a lot, since i modified previous flake implementations with added dependencies and they do not work correctly. (altough im not that proficient at nix yet)
 
> A flake.nix file like neovim has could definitely improve things, especially for those who wish to contribute to godot.

I definetly agree, main problem is maintaining that flake , if the devs do not use it, it could become a burden more than a blessing.

>But yeah a good part of my point was, it wouldnt take a lot. a flake.nix file that outputs a normal and debug build of the package, and a dev shell with the build tools would be plenty sufficient. And it doesnt even need to be at root level of the repo, we can specify the directory of the flake. basically just 2 files (flake.nix and flake.lock) somewhere is all thats required

Does that affect anything in practice? like for example when using it as an input

**Reply by BirdeeHub** [reply]
godot.url = "github:godotengine/godot?dir=the/path"
